pkg/sql/catalog/seqexpr/sequence.go

```diff
@@ -187,3 +190,123 @@ func ReplaceSequenceNamesWithIDs(
+// (`allUsedSeqNames`). This set is expected to contain names of all sequences used
+// in `expr`, so all we need to do is to match each by-name seq reference in `expr`
+// to one entry in `allUsedSeqNames`.
+func UpgradeSequenceReferenceInExpr(
+	expr *string, allUsedSeqNames map[descpb.ID]*tree.TableName,
+) (hasUpgraded bool, err error) {
+	// Find all sequence references in `expr`.
+	parsedExpr, err := parser.ParseExpr(*expr)
+	if err != nil {
+		return hasUpgraded, err
+	}
+	seqRefs, err := GetUsedSequences(parsedExpr)
+	if err != nil {
+		return hasUpgraded, err
+	}
+
+	// Construct the key mapping from seq-by-name-reference to their IDs.
+	seqByNameRefToID := make(map[string]int64)
+	for _, seqIdentifier := range seqRefs {
+		if seqIdentifier.IsByID() {
+			continue
+		}
+
+		parsedSeqName, err := parser.ParseTableName(seqIdentifier.SeqName)
+		if err != nil {
+			return hasUpgraded, err
+		}
+		seqByNameRefInTableName := parsedSeqName.ToTableName()
+
+		// Pairing: find out which sequence name in `allUsedSeqNames` matches
+		// `seqByNameRefInTableName` so we know the ID of this seq identifier.
+		idOfSeqIdentifier, err := findUniqueBestMatchingForTableName(allUsedSeqNames, seqByNameRefInTableName)
+		if err != nil {
+			return hasUpgraded, err
+		}
+
+		seqByNameRefToID[seqIdentifier.SeqName] = int64(idOfSeqIdentifier)
+	}
+
+	// With this name-to-ID mapping, we can upgrade `expr`.
+	newExpr, err := ReplaceSequenceNamesWithIDs(parsedExpr, seqByNameRefToID)
+	if err != nil {
+		return hasUpgraded, err
+	}
+
+	// Modify `expr` in place, if any upgrade.
+	if *expr != tree.Serialize(newExpr) {
+		hasUpgraded = true
+		*expr = tree.Serialize(newExpr)
+	}
+
+	return hasUpgraded, nil
+}
+
+// findUniqueBestMatchingForTableName picks the "best-matching" name from `allTableNamesByID` for `targetTableName`.
+// The best-matching name is the one that matches all parts of `targetTableName`, if that part exists in both names.
+//
+// Example 1:
+// 		allTableNamesByID = {23 : 'db.sc1.t', 25 : 'db.sc2.t'}
+//		tableName = 'sc2.t'
+//		return = 25 (because `db.sc2.t` best-matches `sc2.t`)
+// Example 2:
+// 		allTableNamesByID = {23 : 'db.sc1.t', 25 : 'sc2.t'}
+//		tableName = 'sc2.t'
+//		return = 25 (because `sc2.t` best-matches `sc2.t`)
+// Example 3:
+// 		allTableNamesByID = {23 : 'db.sc1.t', 25 : 'sc2.t'}
+//		tableName = 'db.sc2.t'
+//		return = 25 (because `sc2.t` best-matches `db.sc2.t`)
+//
+// It returns a non-nill error if `tableName` does not uniquely match a name in `allTableNamesByID`.
+//
+// Example 4:
+// 		allTableNamesByID = {23 : 'sc1.t', 25 : 'sc2.t'}
+//		tableName = 't'
+//		return = non-nil error (because both 'sc1.t' and 'sc2.t' are equally good matches
+//	 			for 't' and we cannot decide,	i.e., >1 valid candidates left.)
+// Example 5:
+// 		allTableNamesByID = {23 : 'sc1.t', 25 : 'sc2.t'}
+//		tableName = 't2'
+//		return = non-nil error (because neither 'sc1.t' nor 'sc2.t' matches 't2', that is, 0 valid candidate left)
```

Comment:
Can you add tests to cover these cases in the commentary? In particular I'd like to see a rename fail because there's a sequence that's still referenced by name somewhere because the name is ambiguous, preventing it from being upgraded to a reference by ID.